### PR TITLE
Implement new publishing mode for local metrics

### PR DESCRIPTION
**Pull request by Neverlord**
@sethhall these are the Broker-side adjustments for the telemetry framework we've been talking about the other day.

The gist is that you can now set `BROKER_METRICS_TOPIC` to spin up a small worker that periodically collects all local metrics and publishes them at the given topic. For example, you could have worker-1 publish to `/zeek/metrics/worker-1`, worker-2 to `/zeek/metrics/worker-2`, and so on. Then you just subscribe to `/zeek/metrics` at a central point and wait for the updates to tickle in. Currently, the default interval is 1s, which could prove a bit too aggressive in a somewhat large-ish cluster.

Currently only supports counters and gauges, adding histograms is straightforward though.

At the central point, you'll get a `vector` with one metric per "line". Each line is a vector again with five elements: prefix, name, type, labels, value. Looks like this if you're just printing it:

```
[/my/metric/server-1, ((caf.system, rejected-messages, counter, {}, 0), (caf.system, processed-messages, counter, {}, 122), (caf.system, running-actors, gauge, {}, 5), (caf.system, queued-messages, gauge, {}, 0))]
```

There's also a filter option that lets you specify what prefixes you're interested in.

### CODE REVIEW

**Review by jsiwek — COMMENTED** [review]
Looks nice, thanks.  Is this supposed to wait for feedback from @sethhall specifically before merging ?  I only had a few thoughts below:

> Currently only supports counters and gauges, adding histograms is straightforward though.

Not sure if that's waiting for feedback before doing the work, but seems fine to add histogram support for completeness (either as part of this PR, or even later when time permits would be ok by me).

> Currently, the default interval is 1s, which could prove a bit too aggressive in a somewhat large-ish cluster.

If those with smaller clusters would still generally be content with lower frequency results, I'd say bump the suggested default up a bit so it may work better on larger clusters.  No idea what that would be though and no strong feeling about it since it's ultimately tunable either way.

File: src/detail/telemetry_scraper.cc
Comment (Neverlord) [reply src/detail/telemetry_scraper.cc:79]: Yeah, it does look a bit convoluted at a first glance. Why not just use `delayed_send`? So here's my reasoning to do it this way.

When using `delayed_send` (or manually scheduling based on `now` + some interval), you'll always have some drift over time because the messages never arrive at precisely the scheduled time and the time for processing the timeout itself is also subject to variations (unpredictable interrupts, shifting cache hit rates, and so on). Say our first scrape was at precisely 09:00:00 and your interval is 5s. With this code, Broker tries to scrape at ...:05, ...:10, and so on. If one scrape gets delayed (because of scheduling artifacts or load peaks), then at least the next one should be back on time.

A drift over time is no big deal on a single system. However, with multiple distributed workers and each drifting slightly different over time, I think it's worth the bit of extra bookkeeping to try and have each worker publish its data on a somewhat predictable schedule.

**Comment by Neverlord** [comment]
> Looks nice, thanks. Is this supposed to wait for feedback from @sethhall specifically before merging ?

If we don't actually end up using this feature in the new telemetry framework then the question is whether it's still a valuable feature to have on its own. I can see it being useful for debugging performance issues since it gives access to all the telemetry data in one place, but maybe we should ask people that actually debugged performance issues on live system whether it's something that they'd use. Maybe we first need a way to feed the collected data to Prometheus first, though. Not sure if just looking at some data on the command line (as opposed to having everything in Prometheus/Grafana) is enough of a feature right now.

> >  Currently only supports counters and gauges, adding histograms is straightforward though.
>
> Not sure if that's waiting for feedback before doing the work, but seems fine to add histogram support for completeness (either as part of this PR, or even later when time permits would be ok by me).

I mostly thought it'd be worth waiting on the first round of feedback before making the branch feature complete in case there are some structural changes necessary. I'll add the histogram support to this PR.

**Comment by Neverlord** [comment]
FWIW, CI failed only because of some resource restrictions:

Screen Shot 2021-04-22 at 07 55 02

**Comment by Neverlord** [comment]
At least @0xxon has immediate use for the Prometheus export of remote metrics, so I've started working on that. :)

**Comment by Neverlord** [comment]
Ok, I've implemented publishing metrics to topics as well as collecting remote metrics. Also made some tweaks to the encoding of metrics to `broker::data` to make collecting metrics easier.

From a user perspective, this is how it works when using the environment variables:

- Setting `BROKER_METRICS_PORT` spins up a background thread that waits for Prometheus queries (via HTTP). Unlike the default CAF export, Broker adds an implicit new label dimension to all metrics: `endpoint`. When not setting anything else, this is `"localhost"` by default.
- Setting `BROKER_METRICS_ENDPOINT_NAME` forces a value for the `endpoint` label dimension.
- For collecting remote metrics, users can pass a comma-separated list of topics via `BROKER_METRICS_IMPORT_TOPICS`.
- For making local metrics available to others, users can set `BROKER_METRICS_EXPORT_TOPIC` to cause Broker to periodically publish local metrics on given topic.

For my local testing, I've ran the server like this:

```sh
BROKER_METRICS_PORT=4040 \
BROKER_METRICS_ENDPOINT_NAME=server \
BROKER_METRICS_IMPORT_TOPICS=/zeek/metrics \
./build/release/bin/broker-benchmark --server :1234
```

Prometheus is configured to collect metrics from `localhost:4040`. The metrics for the server process then show up with `endpoint="server"` in Prometheus/Grafana.

For the client, we only need to set the topic:

```sh
BROKER_METRICS_EXPORT_TOPIC=/zeek/metrics/worker-1 \
./build/release/bin/broker-benchmark -r0 localhost:1234
```

If Broker finds a topic in `BROKER_METRICS_EXPORT_TOPIC` then it sets the endpoint name to the suffix. Hence, metrics for the client process show up with `endpoint="worker-1"` in Prometheus/Grafana. Users can of course still override the endpoint name via `BROKER_METRICS_ENDPOINT_NAME`.

There are a couple of additional variables for fine-tuning the process:

- `BROKER_METRICS_EXPORT_INTERVAL` configures how frequent clients push their local metrics to their topic (default: 1s).
- `BROKER_METRICS_EXPORT_PREFIXES` optionally may store a list of metric prefixes for publishing only a subset of the local metrics.

Side note: the CAF submodule points to a recent HEAD version of CAF for some minor API additions that I needed to re-use the Prometheus writer. 0.18.3 is going to release soon, but we can delay merging the PR until having the official tag if that's an issue.

**Comment by jsiwek** [comment]
> Is there a way to set BROKER_METRICS_EXPORT_TOPIC, BROKER_METRICS_ENDPOINT_NAME, BROKER_METRICS_PORT, and BROKER_METRICS_IMPORT_TOPICS programmatically from Zeek scripts? It feels like having that ability would offer a nicer user-interface in many cases.

No real objection about whether to add explicit tuning options for those in Zeek, but for a way of currently using `setenv()` from scripts, you can try something like this:

```zeek
function init_env(): bool
	{
	local rval = setenv("MY_ENV_VAR", "this works");
	return rval;
	}

global init_env_result = init_env();

event zeek_init()
	{
	if ( init_env_result )
		print getenv("MY_ENV_VAR");
	else
		Reporter::fatal("failed to init_env()");
	}
```

The trick is to use `setenv()` within a global-variable initialization rather than within `zeek_init()` so that's it's available at an earlier time, and seen by Broker's setup logic.

**Comment by Neverlord** [comment]
@0xxon is Jon's suggestion working for you or do you still need the parameters to be run-time configurable?

**Comment by rsmmr** [comment]
`setenv()` is nice work-around but yeah, script-level options seem the way to go. It would be fine I think if they weren't changeable at runtime, so just global `const` with `&redef` would work if that's easier than making them real `option` that can update any time.

**Comment by Neverlord** [comment]
> `setenv()` is nice work-around but yeah, script-level options seem the way to go. It would be fine I think if they weren't changeable at runtime, so just global `const` with `&redef` would work if that's easier than making them real `option` that can update any time.

Ok, then I'll make it changeable at run-time. Btw, `const &redef` vs. `option` doesn't make a difference on the Broker side so that's up to preference later. 🙂

**Comment by 0xxon** [comment]
> The trick is to use `setenv()` within a global-variable initialization rather than within `zeek_init()` so that's it's available at an earlier time, and seen by Broker's setup logic.

That is a really cute workaround that I would never have thought of. I am going to use that to deploy a running setup using this branch... hopefully within the next few days. Let me see if I encounter any more problems when actually trying it :)

**Comment by rsmmr** [comment]
> > `setenv()` is nice work-around but yeah, script-level options seem the way to go. It would be fine I think if they weren't changeable at runtime, so just global `const` with `&redef` would work if that's easier than making them real `option` that can update any time.
> 
> Ok, then I'll make it changeable at run-time. Btw, `const &redef` vs. `option` doesn't make a difference on the Broker side so that's up to preference later. 🙂

Then `option` is preferred, just note that one can change those during runtime at any time, wouldn't Broker need to react to that, say if I changed the port? (This is advanced stuff, we have hooks for reacting to option changes if additional logic needs to trigger at that point; was just trying to keep it easy :-)

**Comment by Neverlord** [comment]
> Then `option` is preferred, just note that one can change those during runtime at any time, wouldn't Broker need to react to that, say if I changed the port? (This is advanced stuff, we have hooks for reacting to option changes if additional logic needs to trigger at that point; was just trying to keep it easy :-)

Yeah, changing the port at runtime would be a challenge. Because with a port, Broker spins up a background thread for doing the socket I/O and otherwise Broker just spins up an actor for doing the publishing. Changing this at runtime would get tricky.

@0xxon I left the port out for now, but everything else can be changed at runtime now. Is having the port still bound via environment variable acceptable? I'd imagine only one node in the system should need to spin up the Prometheus exporter for handling the HTTP requests. Everything else is configurable in script land now. See https://github.com/zeek/zeek/pull/1562.

**Comment by jsiwek** [comment]
> I left the port out for now, but everything else can be changed at runtime now. Is having the port still bound via environment variable acceptable?

I don't have input on whether changing port at *run-time* is desired (assuming it's not from prior discussion), but if you still want to provide a *startup-time* alternate to the environment variable in form of a Zeek-side API, think you can add something like this to the `Broker` module's `export` section:

```zeek
const metrics_port = "" &redef;
```

Then set the `broker.metrics.port` config option based on the value of that `Broker::metric_port` constant from within `zeek::Broker::Manager::InitPostScript()`, at the point where it's creating the `broker::endpoint`.

**Comment by rsmmr** [comment]
> I don't have input on whether changing port at _run-time_ is desired (assuming it's not from prior discussion), but if you still want to provide a _startup-time_ alternate to the environment variable in form of a Zeek-side API, 

Yep, agree, we don't need the dynamic change at run-time, but should still offer the redefable constant for customization.

**Review by jsiwek — APPROVED** [review]

**Comment by Neverlord** [comment]
@jsiwek thanks for the pointers! I followed your suggestions on the Zeek PR to make the port also configurable via script. Luckily didn't require any changes on the Broker side.

**Pull request merged**
